gateway/mdb/models/machine.py

```diff
@@ -1,66 +1,100 @@
-
+import datetime
+import enum
+import hashlib
 import random
 import string
-from django.db import models
+from typing import Self
 
+from django.conf import settings
 from django.core.validators import RegexValidator
+from django.db import models
 
 from gateway.rules import require_not_in_production
-
-from mdb.models.room import Room
 from mdb.models.mgroup import MachineGroup
+from mdb.models.room import Room
 
-import hashlib
-
-REGEX_MAC    = r"[0-9a-fA-F]{2}(:[0-9a-fA-F]{2}){5}"
+REGEX_MAC = r"[0-9a-fA-F]{2}(:[0-9a-fA-F]{2}){5}"
 REGEX_SECRET = r"[0-9a-fA-F]{64}"
 
-class Machine (models.Model):
-    host = models.CharField(
-        max_length = 32, 
-        unique = True,
-        verbose_name = "Host Name"
-    )
+
+@enum.unique
+class ConnectionStatus(enum.Enum):
+    Connected = enum.auto()
+    Unknown = enum.auto()
+    Disconnected = enum.auto()
+
+    @classmethod
+    def default() -> Self:
+        ConnectionStatus.Unknown
+
+    @classmethod
+    def from_time_elapsed(time: datetime.timedelta) -> Self:
+        assert isinstance(time, datetime.timedelta)
+
+        if time <= settings.PING_INTERVAL + settings.PING_INTERVAL_TOLERANCE:
+            return Self.Connected
+        elif time <= 2 * settings.PING_INTERVAL + settings.PING_INTERVAL_TOLERANCE:
+            return Self.Unknown
+        else:
+            return Self.Disconnected
+
+
+class Machine(models.Model):
+    host = models.CharField(max_length=32, unique=True, verbose_name="Host Name")
     mac = models.CharField(
-        max_length = 17,
-        unique = True,
-        verbose_name = "MAC Address",
-        validators = [ RegexValidator(regex=REGEX_MAC, message="Enter valid MAC Address") ]
+        max_length=17,
+        unique=True,
+        verbose_name="MAC Address",
+        validators=[RegexValidator(regex=REGEX_MAC, message="Enter valid MAC Address")],
     )
     secret = models.CharField(
-        max_length = 64,
-        verbose_name = "Secret",
-        validators = [ RegexValidator(regex=REGEX_SECRET, message="Invalid Secret") ],
-        null = True
+        max_length=64,
+        verbose_name="Secret",
+        validators=[RegexValidator(regex=REGEX_SECRET, message="Invalid Secret")],
+        null=True,
+    )
+
+    room = models.ForeignKey(Room, on_delete=models.PROTECT, verbose_name="Room")
+    group = models.ForeignKey(
+        MachineGroup, on_delete=models.PROTECT, verbose_name="Group"
     )
 
-    room  = models.ForeignKey( Room,         on_delete=models.PROTECT, verbose_name="Room" )
-    group = models.ForeignKey( MachineGroup, on_delete=models.PROTECT, verbose_name="Group" )
+    last_ping = models.DateTimeField(verbose_name="Last ping", default=None)

```

Comment:
Last ping should have a default value of now (makes sens because the mdb was pinged to create the record), and otherwise it fails 20 tests from the CI/CD.